geospace-lib/src/main/scala/com/socrata/geospace/lib/shapefile/ZipUtils.scala

```diff
@@ -73,6 +78,73 @@ class TemporaryZip(compressed: Array[Byte]) extends Closeable with Logging {
     FileUtils.deleteDirectory(contents)
     logger.info("Deleted temporary extracted shapefile contents")
     archive.delete()
+    tmpDir match{
+      case Some(x) => x.delete()
+      case None => Unit
+    }
     logger.info("Deleted temporary shapefile zip")
   }
+
 }
+
+
+/** Saves the contents of a zip file in a single flattened directory
+  * and cleans up all temporary files afterwards.
+  *
+  * @constructor create a new temporary zip given the zip file as a byte array.
+  * @param compressed byte array representation of the zip file
+  */
+case class ZipFromArray(compressed: Array[Byte]) extends AbstractZip(None) {
+  require(Option(compressed).filter(_.nonEmpty).isDefined, "Null or empty zip file")
+
+
+  def getArchive = {
+    for {
+      in <- managed(new ByteArrayInputStream(compressed))
+      out <- managed(new FileOutputStream(tmpFile))
+    } {
+      IOUtils.copy(in, out)
+      logger.info("Temporarily copied shapefile zip to {}", tmpFile.getAbsolutePath)
+    }
+
+    tmpFile
+  }
+}
+
+/** Saves the contents of a zip file in a single flattened directory
+  * and cleans up all temporary files afterwards.
+  *
+  * @constructor create a new temporary zip given the zip file inputstream.
+  * @param compressed inputstream representation of the zip file
+  * @param tmpDir temp directory where to place the zip file.
+  */
+case class ZipFromStream(compressed: InputStream, tmpDir: Option[File]) extends AbstractZip(tmpDir) {
+  require(Option(compressed).filter(_.available >=0).isDefined, "Null or empty zip file")
+
+  def getArchive = {
+    val ops = new FileOutputStream(tmpFile)
+    val buff= new Array[Byte](1024)
+
+    @tailrec
+    def write(): Unit = {
+      val read = compressed.read(buff)
+      if (read != -1){
+        ops.write(buff,0,read)
+        write()
+      }
+    }
+
+    write()
+    ops.flush()
+    ops.close()
```

Comment:
Use a for comprehension or try/finally here so the stream gets closed correctly even in the exception case.